[poro] step-5 Controller 단위 테스트, 인증 및 게시글 작성/수정/삭제 기능 구현(요청사항 반영)

Gwonwoo-Nam opened the pull request:
## 학습 내용
- 피드백으로 알려주신 @SpringBootTest외에 다른 어노테이션에 대해서 알아보다가, @WebMvcTest에 대해서 알게 되었습니다. 위 어노테이션을 사용하면 Controller와 관련된 Bean들만 스캔되며, @TestConfiguration와 @ContextConfiguration으로 원하는 환경으로 쉽게 설정할 수 있었습니다. 컨트롤러를 테스트할 때 의존성을 가지는 다른 객체에 대해서 어떻게 의존성을 분리해서 테스트를 할 지 고민을 했었는데 MockBean으로 예상되는 행동을 지정해주어 컨트롤러의 작동만을 테스트할 수 있다는 것을 알게되어 이번 기회에 적용해보았습니다.
- 데이터베이스를 테스트할 때도 마찬가지로, @SpringBootTest가 아닌 어노테이션을 찾아보았습니다. 그 중에서 저는 @JdbcTest를 사용했습니다.
- Spring Data JDBC를 사용하지 않고 오직 DataSource를 필요로하는 테스트에서 사용된다고하는데, @JdbcTest는 in-memory embedded database가 설정되어 테스트를 위한 JdbcTemplate을 생성하고 일반적인 component 스캔을 하지 않습니다. 기본적인 @트랜잭셔널을 기본으로 적용해줍니다.

## 고민한 내용
- 구현 요구사항에 따라 코드를 변경하면 단위 테스트가 전부 어긋나버려 함께 수정해야하는데 항상 고민입니다. 현재 단계에서는 테스트의 필요성은 알고있지만 테스트코드로 인해 생산성이 꽤 저하되는 것 같습니다. 유연하지 못한 설계를 가지고 있기 때문에 유지보수가 어려운건지, 그렇다면 어떻게 고쳐봐야할지 고민이 많이 됩니다.

Gwonwoo-Nam committed:
refactor(step-5):피드백 반영하여 불필요한 코드 정리

- 미사용 코드 제거, User의 index setter 제거
- login 요청 URL login/validate에서 login으로 통합

Gwonwoo-Nam committed:
test(step-5):mockMvc로 UserController Test 구현

- DataSource가 DI되지 않아 ContextConfiguration으로 테스트 Config 별도 설정

Gwonwoo-Nam committed:
refactor(step-5):test도 repository, service로 패키지 분리

- User의 public method에 대한 테스트도 생성, 구현 예정

Gwonwoo-Nam committed:
test(step-5):피드백 반영하여 SpringBootTest를 JdbcTest 어노테이션으로 변경, 초기화 코드 추가

- In Memory DB 초기화 쿼리 추가(create table)

Gwonwoo-Nam committed:
feat(step-5):인증 및 권한 부여 기능 구현

- 로그인하지 않은 사용자는 게시글 목록만 볼 수 있다.
- HandlerInterceptor인 login 인터셉터를 구현하여 인증 기능 구현

Gwonwoo-Nam committed:
feat(step-5):권한 부여에 따른 게시글 작성 기능 추가, 게시글 수정 기능 구현

- Article 타입이 user를 참조, 코드 기능 분리를 위해 ArticleDto 추가 구현
- 게시글 작성에서 글쓴이 입력 필드 삭제, 세션의 ID를 조회해 Article에 User Id를 부여하는 로직
- 아이디가 다른 경우 수정하지 못하도록 에러 페이지로 리다이렉션
- 수정 기능 추가 구현 예정

Gwonwoo-Nam committed:
feat(step-5):게시글 수정하기 기능 구현

- 뷰 modification_form를 만들어 수정 Put Request로 전송
- Article의 ID가 일치하는 레코드를 업데이트한다.

Gwonwoo-Nam committed:
feat(step-5):게시글 삭제 기능 구현

- 로그인 사용자가 다른 경우에는 에러페이지로 리다이렉트

honux77 reviewed (changes requested):
머지 방지

honux77 commented:
구현 요구사항에 따라 코드를 변경하면 단위 테스트가 전부 어긋나버려 함께 수정해야하는데 항상 고민입니다. 현재 단계에서는 테스트의 필요성은 알고있지만 테스트코드로 인해 생산성이 꽤 저하되는 것 같습니다. 유연하지 못한 설계를 가지고 있기 때문에 유지보수가 어려운건지, 그렇다면 어떻게 고쳐봐야할지 고민이 많이 됩니다. --> 답을 쓰려다 저도 고민이 많아져서... 요거는 나중에 수업때 의논해 봅시다. 

https://www.youtube.com/watch?v=rR4n-0KYeKQ

honux77 commented on src/main/java/kr/codesqaud/cafe/AutoAppConfig.java:
```diff
@@ -41,17 +50,23 @@ public QnaService qnaService() {
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addInterceptor(loginInterceptor)
+                .addPathPatterns("/**")
```
로그인 인터셉터가 모든 URL에 대응되는 것보다는 필요한 URL만 추가하는게 좋지 않을까요?

Gwonwoo-Nam replied:
해당사항 반영했습니다. css, js 등 static 파일을 포함시키지 않아도 되어 훨씬 간결해졌습니다.

honux77 commented on src/main/java/kr/codesqaud/cafe/model/ArticleDto.java:
```diff
@@ -0,0 +1,46 @@
+        return creationTime;
+    }
+
+    public String getFormattedCreationTime() {
+        return creationTime.format(
+                DateTimeFormatter.ofPattern("yyyy-MM-dd hh:mm:ss"));
+    }
```
DTO의 용도는 말 그대로 Data Transfer인데 이 메소드가 왜 필요하게 되었는지 궁금하네요.

Gwonwoo-Nam replied:
말씀해주신대로 DTO가 로직을 가지기보다는 format을 처리한 후 DTO로 넘겨 DTO는 데이터를 전달해주는 역할만을 수행하도록 수정했습니다.

honux77 commented on src/main/java/kr/codesqaud/cafe/exceptions/ArticleInfoException.java:
```diff
@@ -0,0 +1,22 @@
+package kr.codesqaud.cafe.exceptions;
+
+public class ArticleInfoException extends RuntimeException {
```
RuntimeException을 상속한 이유가 있을까요?

Gwonwoo-Nam replied:
사용자의 실행 중에 발생하는 예외라 생각하여 RuntimeException을 상속했는데 CheckedException이 컴파일러의 도움을 받을 수 있기 때문에 더 좋은 것 같습니다. 아직 Exception에 대해 잘 모르는 것 같아 더 공부해보겠습니다.

honux77 commented on src/main/java/kr/codesqaud/cafe/service/impl/JoinServiceImpl.java:
Join 서비스라는 서비스 이름은 적절하지 않다고 생각합니다.
비지니스 로직을 서비스 계층에서 구현하는데 User나 Article등 특정 도메인에 속하도록 이름을 변경하는게 좋지 않을까요?

honux77 commented on src/main/java/kr/codesqaud/cafe/service/impl/QnaServiceImpl.java:
Join 서비스와 Qna 서비스에 인터페이스가 꼭 있어야 되는 건지 고민해 보시면 좋겠어요. 구현체에도 Impl 보다 명확한 이름이 있으면 좋겠습니다.

Gwonwoo-Nam replied:
지적해주신대로 서비스가 있을 필요성을 저도 못느껴서 삭제하고 레포지토리로 바로 연결했습니다.

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -1,26 +1,38 @@
+    public String articlePost(@RequestParam String title, @RequestParam String contents, HttpSession httpSession) {
+        String userId = (String)httpSession.getAttribute("sessionedUser");
+
+        qnaService.postQna(new Article(joinService.lookupUser(userId), title, contents));
```
UserRepo의 findById가 <Optional> 이었으면 좋았을 것 같구요. lookupUser가 join 서비스에 있어야 하는지도 고민해 보면 좋겠어요. 

그리고 세션 유저와 사용자 유저가 달랐을 경우의 예외처리가 명확하지 않아 보입니다.

Gwonwoo-Nam replied:
URL에 사용자 유저에 대한 정보가 없고(user/create) 글 게시 시에는 현재 세션의 유저 이름을 가져오기 때문에 예외가 발생할 경우가 없는것같은데 혹시 제가 생각하지 못한 예외 경우가 있을까요?

honux77 replied:
앗 그렇네요. 이건 제가 잘못 생각했어요. 그리고 RuntimeException을 상속받는 것도 장점이 많군요. 검색해 보니 RuntimeException을 더 많이 사용하는 것 같아요.

honux77 reviewed (changes requested):
특별한 이유가 없었다면 커스텀 예외 처리를 Exception 클래스를 상속받아서 만들고 
예외처리를 약간 명확하게 하면 좋을 것 같습니다.

Gwonwoo-Nam committed:
refactor(step-5):Service 계층 삭제, Controller에서 바로 Repository로 연결

- Service 계층이 하는 역할이 없다고 생각되어 삭제, Repository와 Controller를 연결

Gwonwoo-Nam committed:
refactor(step-5):ArticleDto의 Getter의 날짜 formatting 로직 삭제, 예외 RuntimeException에서 Exception을 상속받도록 변경

- Dto에서 날짜를 formatting하기보다 Respository에서 처리한 후 Data Transfer의 역할만 맡도록 분리
- RuntimeException을 상속하는 Custom 예외 클래스를 Exception으로 컴파일러에 의해 체크하도록 변경

Gwonwoo-Nam committed:
chore(step-5):import 삭제 및 Configuration 파일 설정 서비스 계층 삭제에 따라 변경

Gwonwoo-Nam commented:
피드백 주신 사항 반영했습니다. 피드백 주신 내용 중 한 가지 궁금한 점은 article post 시에 현재 세션으로부터 사용자 이름을 받아와 Repository에서 유저를 조회하고,  세션 유저의 이름으로 글을 작성하는데 이러한 경우에도 예외가 발생할 수 있는 건지 궁금합니다! 감사합니다.

honux77 commented:
예외처리는 괜히 고친 것 같아요 ㅎㅎ 죄송합니다. 요것도 수업때 한 번 생각해 봐요.
